src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs

```diff
@@ -11,108 +11,178 @@
 {
+    /// <summary>
+    /// A test fixture for sharing pushed packages between tests. Since packages take quite a while (minutes) to make
+    /// their way through the V3 pipeline, packages should be pushed all at once and shared between tests, if possible.
+    /// </summary>
     public class PushedPackagesFixture : IDisposable
     {
-        private readonly object _packagesLock;
-        private readonly IDictionary<PackageType, PushedPackage> _packages;
+        private static readonly HashSet<PackageType> SemVer2PackageTypes = new HashSet<PackageType>
+        {
+            PackageType.SemVer2Prerelease,
+        };
+
+        private readonly SemaphoreSlim _pushLock;
+        private readonly object _packagesLock = new object();
+        private readonly IDictionary<PackageType, Package> _packages;
         private readonly Clients _clients;
+        private readonly TestSettings _testSettings;
 
         public PushedPackagesFixture()
         {
             _clients = Clients.Initialize();
-            _packagesLock = new object();
-            _packages = new Dictionary<PackageType, PushedPackage>();
+            _testSettings = TestSettings.CreateFromEnvironment();
+            _pushLock = new SemaphoreSlim(initialCount: 1);
+            _packages = new Dictionary<PackageType, Package>();
         }
         
-        public async Task<Package> PushAsync(PackageType packageType, ITestOutputHelper logger)
+        public async Task<Package> PushAsync(PackageType requestedPackageType, ITestOutputHelper logger)
         {
-            PushedPackage pushedPackage;
-            lock (_packagesLock)
+            var pushedPackage = GetPushedPackage(requestedPackageType, logger);
+            if (pushedPackage != null)
             {
-                if (!_packages.TryGetValue(packageType, out pushedPackage))
-                {
-                    var package = CreatePackage(packageType);
-                    pushedPackage = new PushedPackage(package);
-                    _packages[packageType] = pushedPackage;
-                }
+                return pushedPackage;
             }
 
-            if (pushedPackage.Pushed)
-            {
-                logger.WriteLine($"Package {pushedPackage} push has already been pushed.");
-                return pushedPackage.Package;
-            }
+            return await PushUnpushedPackagesAsync(requestedPackageType, logger);
+        }
 
-            var acquired = await pushedPackage.PushLock.WaitAsync(0);
+        private async Task<Package> PushUnpushedPackagesAsync(PackageType requestedPackageType, ITestOutputHelper logger)
+        {
+            var acquired = await _pushLock.WaitAsync(0);
             try
             {
                 if (!acquired)
                 {
-                    logger.WriteLine($"Another test is already pushing {pushedPackage}. Waiting.");
-                    await pushedPackage.PushLock.WaitAsync();
-                    acquired = true;
```

Comment:
I'm a little confused. Where in this scope is acquired set to true? the line setting it to true explicitly appears to have been removed with your changeset?